cpp/src/groupby/common/m2_var_std.cu

```diff
@@ -0,0 +1,196 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "m2_var_std.hpp"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf::groupby::detail {
+
+namespace {
+
+template <typename Source>
+__device__ constexpr bool is_m2_supported()
+{
+  return is_numeric<Source>() && !is_fixed_point<Source>();
+}
+
+struct m2_functor {
+  template <typename Source, typename... Args>
+  void operator()(Args...)  //
+    requires(!is_m2_supported<Source>())
+  {
+    CUDF_FAIL("Invalid source type for M2 aggregation.");
+  }
+
+  template <typename Target, typename SumSqrType, typename SumType, typename CountType>
+  void evaluate(Target* target,
+                SumSqrType const* sum_sqr,
+                SumType const* sum,
+                CountType const* count,
+                size_type size,
+                rmm::cuda_stream_view stream) const noexcept
+  {
+    thrust::tabulate(rmm::exec_policy_nosync(stream),
+                     target,
+                     target + size,
+                     [sum_sqr, sum, count] __device__(size_type const idx) {
+                       auto const group_count = count[idx];
+                       if (group_count == 0) { return Target{}; }
+                       auto const group_sum_sqr = static_cast<Target>(sum_sqr[idx]);
+                       auto const group_sum     = static_cast<Target>(sum[idx]);
+                       auto const result = group_sum_sqr - group_sum * group_sum / group_count;
+                       return result;
+                     });
+  }
+
+  template <typename Source>
+  void operator()(mutable_column_view const& target,
+                  column_view const& sum_sqr,
+                  column_view const& sum,
+                  column_view const& count,
+                  rmm::cuda_stream_view stream) const noexcept  //
+    requires(is_m2_supported<Source>())
+  {
+    using Target     = cudf::detail::target_type_t<Source, aggregation::M2>;
+    using SumSqrType = cudf::detail::target_type_t<Source, aggregation::SUM_OF_SQUARES>;
+    using SumType    = cudf::detail::target_type_t<Source, aggregation::SUM>;
+    using CountType  = cudf::detail::target_type_t<Source, aggregation::COUNT_VALID>;
+
+    // Separate the implementation into another function, which has fewer instantiations since
+    // the data types (target/sum/count etc) are mostly the same.
```

Comment:
`m2_functor` is used only once in the code. how is this reducing instantiations?

Reply:
This is the dispatch functor and this function will be instantiated many times for various data type (numeric not fixed point). On the other hand, the function `evaluate` is instantiated less, with sum type is  either `int64` or  `float/double`.